File: internal/executor/charset_compatibility_test.go
package executor

import (
	"encoding/base64"
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
)

type charsetCompatibilityManifest struct {
	SchemaVersion string                     `json:"schemaVersion"`
	Cases         []charsetCompatibilityCase `json:"cases"`
}

type charsetCompatibilityCase struct {
	ID            string `json:"id"`
	BytesBase64   string `json:"bytesBase64"`
	ContentType   string `json:"contentType"`
	ExpectedHTML  string `json:"expectedHTML"`
	ExpectedError string `json:"expectedError"`
}

func TestCharsetCompatibilityManifest(t *testing.T) {
	path := filepath.Join("..", "..", "fixtures", "html-compat", "charset-manifest.json")
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatal(err)
	}
	var manifest charsetCompatibilityManifest
	if err := json.Unmarshal(data, &manifest); err != nil {
		t.Fatal(err)
	}
	if manifest.SchemaVersion != "2026-07-15" || len(manifest.Cases) == 0 {
		t.Fatalf("manifest contract = %#v", manifest)
	}
	seen := map[string]bool{}
	for _, fixture := range manifest.Cases {
		fixture := fixture
		t.Run(fixture.ID, func(t *testing.T) {
			if fixture.ID == "" || seen[fixture.ID] {
				t.Fatalf("duplicate or empty fixture ID %q", fixture.ID)
			}
			seen[fixture.ID] = true
			body, err := base64.StdEncoding.DecodeString(fixture.BytesBase64)
			if err != nil {
				t.Fatal(err)
			}
			decoded, err := decodeHTML(body, fixture.ContentType)
			if fixture.ExpectedError != "" {
				assertExecutionErrorCode(t, err, fixture.ExpectedError)
				return
			}
			if err != nil || decoded != fixture.ExpectedHTML {
				t.Fatalf("decoded = %q, error = %v", decoded, err)
			}
		})
	}
}

File: packages/scrape-kdl/test/charset-compatibility.test.mjs
import assert from "node:assert/strict";
import { readFile } from "node:fs/promises";
import test from "node:test";
import { compile, ExecutionError } from "../dist/index.js";

const root = new URL("../../../", import.meta.url);
const source = `extractor "charset-compat" version="2026-07-15" language-version="2026-07-15" {
  source "html" { fetch mode="http" url="https://example.test/" }
  field "title" type="string" required=#true { select "h1"; value "text" }
}`;

test("Go and TypeScript share the pinned charset compatibility manifest", async (t) => {
  const manifest = JSON.parse(await readFile(new URL("fixtures/html-compat/charset-manifest.json", root), "utf8"));
  assert.equal(manifest.schemaVersion, "2026-07-15");
  assert.ok(manifest.cases.length > 0);
  const compiled = await compile({ path: "charset-compat.kdl", data: source });
  assert.ok(compiled.program, JSON.stringify(compiled.diagnostics, null, 2));

  for (const fixture of manifest.cases) {
    await t.test(fixture.id, async () => {
      const body = Uint8Array.from(Buffer.from(fixture.bytesBase64, "base64"));
      const extract = () => compiled.program.extract({}, {
        fetch: async () => new Response(body, { headers: { "content-type": fixture.contentType } }),
      });
      if (fixture.expectedError !== undefined) {
        await assert.rejects(extract(), (error) => error instanceof ExecutionError && error.code === fixture.expectedError);
        return;
      }
      const result = await extract();
      assert.equal(result.value.title, fixture.expectedText);
    });
  }
});
